packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -275,6 +367,105 @@ export default createRule<Options, MessageId>({
       if (assertedArgument != null) {
         traverseNode(assertedArgument, true);
       }
+      if (isArrayMethodCallWithPredicate(context, services, node)) {
+        const predicate = node.arguments.at(0);
+        const arrayNode = (node.callee as TSESTree.MemberExpression).object;
+
+        if (predicate) {
+          checkArrayMethodCallPredicate(predicate, arrayNode);
+        }
+      }
+    }
+
+    /**
+     * Dedicated function to check array method predicate calls. Reports predicate
+     * arguments that don't return a boolean value.
+     *
+     * Ignores the `allow*` options and requires a boolean value.
+     */
+    function checkArrayMethodCallPredicate(
+      predicateNode: TSESTree.CallExpressionArgument,
+      arrayNode: TSESTree.Expression,
+    ): void {
+      const isFunctionExpression = isFunctionExpressionNode(predicateNode);
+
+      // custom message for accidental `async` function expressions
+      if (isFunctionExpression && predicateNode.async) {
+        return context.report({
+          node: predicateNode,
+          messageId: 'predicateCannotBeAsync',
+        });
+      }
+
+      const type = checker.getApparentType(
+        services.getTypeAtLocation(predicateNode),
+      );
+
+      const signatures = isFunctionExpression
```

Comment:
I find it odd that changing this to unconditionally use the `getMatchingArrayPredicateSignatures()` heuristic can make _invalid_ tests fail. I'd think it should only really be able to make _valid_ tests fail. Looking at the failures I get, they all involve omitted parameters. So - is there a bug in the heuristic around omitted parameters?

Surely this should be flagged?

([playground](https://deploy-preview-10106--typescript-eslint.netlify.app/play/#ts=5.5.2&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUAPAXIsIC2ARgKYBOAlIgN4C%2BAUANoCMANLiADYcC6AdMQDcyATzQo45ANxA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs%2BSAIZN6AD2LQUySnyaoMkAO5DoTSODABfEFqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))
```ts
function foo(x: number) {}
[1, null].every(foo);
```

Reply:
Oh, hmmmmmmmmm. Upon closer inspection, I think we're running into a `strictFunctionTypes` problem impacting the assignability API. Idk, this kind of makes me want to avoid the heuristics altogether and stick with simply: if it ain't clearly a boolean according to the `getCallSignatures()`, just report it. I just don't think we want it to be our problem to reimplement what TS should have been doing from the start. WDYT? Also maybe would be good to tag in other team members for opinions on this.

Reply:
That's interesting!

I didn't consider reporting this case, as it's normally a type error:

```ts
function foo(x: number) {}
[1, null].every(foo);
```

However, as you pointed out, this code is valid (typescript-wise) with `strict` mode being turned off. I'm so used to having `strict` mode enabled that I didn't even consider this an option.

I can think of ways to fix this "bug" (checking if the parameter is assignable to the value or the other way around. This will check if they have some overlap), but I'm not sure it's worth the maintenance burden and the potential bugs.

As you suggested, I've updated the PR to be a lot simpler and report on any function that may return a non-boolean.

Thanks for your help with this ❤️